aws/efs/cmd/efs-provisioner/volumemetadata.go

```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"os"
+	"path"
+	"strconv"
+
+	"encoding/json"
+	"github.com/golang/glog"
+	"io/ioutil"
+)
+
+const (
+	metadataFile = ".kube-efs-provisioner-metadata"
+)
+
+type volumeMetadata struct {
+	GID              string `json:"gid"`
+	PVCName          string `json:"pvcName"`
+	PVCNamespace     string `json:"pvcNamespace"`
+	StorageClassName string `json:"storageClassName"`
+}
+
+func (v volumeMetadata) GidAsUInt() (uint32, error) {
+	gid, err := strconv.ParseUint(v.GID, 10, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint32(gid), nil
+}
+
+// writeVolumeMetadata writes a serialized version of the given volumeMetadata object to the given directory.
+func writeVolumeMetadata(dir string, md volumeMetadata) error {
+	mdpath := getMetaDataPath(dir)
+
+	contents, err := json.MarshalIndent(md, "", "  ")
+	if err != nil {
+		glog.Errorf("failed to marshal metadata: %v", err)
+		return err
+	}
+
+	if err := ioutil.WriteFile(mdpath, contents, 0600); err != nil {
+		glog.Errorf("failed to write metadata file %v: %v", mdpath, err)
+		return err
+	}
+
+	return nil
+}
+
+// readVolumeMetadata reads the metadata file in the given directory and returns a *VolumeMetadata with the contents.
+func readVolumeMetadata(dir string) (*volumeMetadata, error) {
```

Comment:
A couple things with this function...

- I would not return a pointer. This function returns a small data carrying object that (I believe) should not be changed after it's returned. That is the perfect case for just returning the struct. It's possible that I'm wrong on this point.
- Even if you expect the files to all be tiny, I would never use `ioutil.ReadFile` (or `ioutil.ReadAll`, etc.) in an important service like this. While there are things like memory limits that should be in place to prevent an issue if someone replaces a metadata file with a 2 terabyte blob of random data, the code to stream the deserialization is very nearly as simple as the `ioutil.ReadFile` version.

I would probably rewrite it something like this...

```go
func readVolumeMetadata(dir string) (volumeMetadata, error) {
	var md volumeMetadata
	mdpath := getMetaDataPath(dir)

	r, err := os.Open(mdpath)
	if os.IsNotExist(err) {
		return md, nil
	} else if err != nil {
		glog.Errorf("failed to read metadata file %v: %v", mdpath, err)
		return md, err
	}
	defer r.Close()

	dec := json.NewDecoder(r)
	if err := dec.Decode(&md); err != nil {
		glog.Errorf("failed to unmarshal %v: %v", mdpath, err)
		return md, err
	}

	return md, nil
}
```